modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java

```diff
@@ -95,17 +101,52 @@ protected final <T> T sync(CompletableFuture<T> fut) {
      * @return Whatever the action returns.
      */
     protected final <T> CompletableFuture<T> withSchemaSync(@Nullable Transaction tx, KvAction<T> action) {
-        // TODO: IGNITE-20106 - retry if our request is rejected by the server due to a changed schema version.
+        return withSchemaSync(tx, null, action);
+    }
 
+    private <T> CompletableFuture<T> withSchemaSync(@Nullable Transaction tx, @Nullable Integer previousSchemaVersion, KvAction<T> action) {
         CompletableFuture<Integer> schemaVersionFuture = tx == null
                 ? schemaVersions.schemaVersionAtNow(tbl.tableId())
                 : schemaVersions.schemaVersionAt(((InternalTransaction) tx).startTimestamp(), tbl.tableId());
 
-        CompletableFuture<T> future = schemaVersionFuture.thenCompose(action::act);
+        CompletableFuture<T> future = schemaVersionFuture
+                .thenCompose(schemaVersion -> {
+                    return action.act(schemaVersion)
+                            .handle((BiFunction<T, Throwable, CompletableFuture<T>>) (res, ex) -> {
+                                if (ex != null && isOrCausedBy(InternalSchemaVersionMismatchException.class, ex)) {
+                                    assert tx == null : "Only for implicit transactions a retry might be requested";
+                                    assert previousSchemaVersion == null || !Objects.equals(schemaVersion, previousSchemaVersion)
+                                            : "Same schema version (" + schemaVersion
+                                                    + ") on a retry: something is wrong, is this caused by the test setup?";
+
+                                    // Repeat.
+                                    return withSchemaSync(tx, schemaVersion, action);
+                                }
+
+                                if (ex != null) {
+                                    return failedFuture(ex);
+                                }
+
+                                return completedFuture(res);
+                            });
+                })
+                .thenCompose(identity());
 
         return convertToPublicFuture(future);
     }
 
+    private static boolean isOrCausedBy(Class<? extends Exception> exceptionClass, @Nullable Throwable ex) {
+        if (ex == null) {
```

Comment:
Can we replace this with `return ex != null && (exceptionClass.isInstance(ex) || isOrCausedBy(exceptionClass, ex.getCause()));` ?